flutter_kernel_transformers/lib/track_widget_constructor_locations.dart

```diff
@@ -0,0 +1,567 @@
+          new VariableGet(variable),
+        ));
+        // TODO(jacobr): add an assert verifying the locationField is not
+        // null. Curently we cannot safely add this assert because we do not
+        // handle Widget classes with optional positional arguments. There are
+        // no Widget classes in the flutter repo with optional positional
+        // arguments but it is possible users could add classes with optional
+        // positional arguments.
+        //
+        // constructor.initializers.add(new AssertInitializer(new AssertStatement(
+        //   new IsExpression(
+        //       new VariableGet(variable), _locationClass.thisType),
+        //   conditionStartOffset: constructor.fileOffset,
+        //   conditionEndOffset: constructor.fileOffset,
+        // )));
+      }
+    }
+
+    // Add named parameters to all constructors.
+    for (Constructor constructor in clazz.constructors) {
+      handleConstructor(constructor);
+    }
+    hierarchy.applyChanges(<Class>[clazz]);
+  }
+
+  /// Transform the given new [libraries].  It is expected that all other
+  /// libraries have already been transformed.
+  ///
+  /// It is safe to call this method each time a new list of libraries is
+  /// updated as part of performing a hot reload.
+  void transform(List<Library> libraries) {
+    // TODO(jacobr): can we do something smarter to reuse the existing
+    // hierarchy from the previous state?
+    hierarchy = new IncrementalClassHierarchy();
+
+    if (libraries.isEmpty) {
+      return;
+    }
+
+    _resolveFlutterClasses(libraries);
+
+    if (_widgetClass == null) {
+      // This application doesn't actually use the package:flutter library.
+      return;
+    }
+
+    final Set<Class> transformedClasses = new Set<Class>.identity();
+    final Set<Library> librariesToTransform = new Set<Library>.identity()
+      ..addAll(libraries);
+
+    for (Library library in libraries) {
+      if (library.isExternal) {
+        continue;
+      }
+      for (Class class_ in library.classes) {
+        _transformWidgetConstructors(
+          librariesToTransform,
+          transformedClasses,
+          class_,
+        );
+      }
+    }
+    // Given the hierarchy we are using and the transforms we are applying,
+    // calling this method probably isn't really necessary.
+    hierarchy.applyChanges(transformedClasses);
+
+    // Transform call sites to pass the location parameter.
+    final _WidgetCallSiteTransformer callsiteTransformer =
+        new _WidgetCallSiteTransformer(
+      hierarchy,
+      widgetClass: _widgetClass,
+      locationClass: _locationClass,
+    );
+
+    for (Library library in libraries) {
+      if (library.isExternal) {
+        continue;
+      }
+      library.transformChildren(callsiteTransformer);
+    }
+  }
+
+  bool _instanceOfWidget(Class clazz) {
+    if (clazz == null) {
+      return false;
+    }
+    return hierarchy.getClassAsInstanceOf(clazz, _widgetClass) != null;
```

Comment:
As pointed above, `hierarchy.isSubtypeOf(clazz, _widgetClass)` may be a bit more readable.

Reply:
done

Reply:
I switched these methods but the code fails because the methods are only on ClosedWorldClassHierarchy not on IncrementalClassHierarchy which I believe I need to use.
I'm happy to refactor my code if the helper method is added to the base ClassHierarchy class.

Reply:
Yes, sorry for that confusion.  ClassHierarchy and related classes are being changed currently.  In short, IncrementalClassHierarchy is going to be deprecated (you may already see the related changes in the current SDK checkout), and I strongly recommend to avoid using it.  If that's possible to use the latest changes from the Dart SDK's master branch, I would suggest the following for now:

- Don't update the class hierarchy, create the new one when the classes are updated instead.
- When needed, create a ClassHierarchy using [the recently added factory](https://github.com/dart-lang/sdk/blob/master/pkg/kernel/lib/class_hierarchy.dart#L20).
- When [CL 35700](https://dart-review.googlesource.com/c/sdk/+/35700) is landed (I hope it's going to be today), ClassHierarchy will have isSubtypeOf, and you will be able to use it here and in other analogous places of the transformer.

We do plan to add a component-based class hierarchy, so that it won't be necessary to recreate it in order to update it.